x-pack/plugins/security_solution/server/lib/detection_engine/routes/rules/perform_bulk_action_route.test.ts

```diff
@@ -94,10 +96,38 @@ describe.each([
       });
       const response = await server.inject(getBulkActionRequest(), context);
 
-      expect(response.status).toEqual(403);
+      expect(response.status).toEqual(500);
+      expect(response.body).toEqual({
+        message:
+          "Failed actions: 1. Rules processed: 1. 'Detect Root/Admin Users': 'mocked validation message'",
+        status_code: 500,
+      });
+    });
+
+    it('process the rest of rules actions if one of action fails', async () => {
+      const failedRuleId = 'fail-rule-id';
+      const failedRuleName = 'Rule that fails';
+      clients.rulesClient.find.mockResolvedValue(
+        getFindResultWithMultiHits({
+          data: [{ ...mockRule, id: failedRuleId, name: failedRuleName }, mockRule, mockRule],
+          total: 3,
+        })
+      );
+
+      (buildMlAuthz as jest.Mock).mockReturnValueOnce({
+        validateRuleType: jest
+          .fn()
+          .mockImplementationOnce(() => ({ valid: false, message: 'mocked validation message' }))
+          .mockImplementationOnce(() => ({ valid: true }))
+          .mockImplementationOnce(() => ({ valid: true })),
+      });
+      const response = await server.inject(getUpdateBulkActionRequest(), context);
+
+      expect(response.status).toEqual(500);
       expect(response.body).toEqual({
-        message: 'mocked validation message',
-        status_code: 403,
+        message:
+          "Failed actions: 1. Rules processed: 3. 'Rule that fails': 'mocked validation message'",
+        status_code: 500,
```

Comment:
Can we add some data about succeeded and failed updates to the response body? I think you had it in some previous commit, and it looked like that:

```ts
      expect(response.body).toEqual({
        errors: [
          {
            error: {
              message: 'mocked validation message',
              statusCode: 403,
            },
            rule: {
              id: mockRule.id,
              name: mockRule.name,
            },
          },
        ],
        failed_rules_count: 1,
        rules_count: 1,
        success: false,
      });
```

I'd suggest something like that:

```ts
      expect(response.body).toEqual({
        status_code: 500,
        message: 'Applying bulk edits partially failed',
        rules: {
          requested: 3,
          processed: 3,
          succeeded: 2,
          failed: 1,
        },
        errors: [
          {
            rule_id: mockRule.id,
            rule_name: mockRule.name,
            error_message: 'mocked validation message',
            error_status_code: 403,
          },
        ],
      });
```

`errors` array should probably be explicitly truncated to a large value like 1000 or 10000, as well as strings like `error_message` and `rule_name`.

Reply:
implemented

Reply:
> number of rules limited to 10000

That's a good point, it could be sufficient. I just wasn't sure that it makes sense to return all the errors especially if there's a large number of them. While 10k is still a large number, it could be okish in this case.

> Do we limit any properties sizes in case of find_rules, where response is expected to be much larger than in this bulk edit errors response?

I don't think so. But `rules/_find` is a completely different endpoint that must return the data as it is. Here we're discussing a potentially large array of errors containing data that could be helpful for debugging, and this data doesn't have to be 100% precise or complete imho.

> to be safe, limiting error_message might be a good idea

👍

Reply:
I've added limiting size of error message.